src/serialize.rs

```diff
@@ -231,6 +231,171 @@ pub fn serialized_length_from_bytes(b: &[u8]) -> io::Result<u64> {
     Ok(f.position())
 }
 
+use crate::sha2::{Digest, Sha256};
+
+fn hash_atom(buf: &[u8]) -> [u8; 32] {
+    let mut ctx = Sha256::new();
+    ctx.update(&[1_u8]);
+    ctx.update(buf);
+    ctx.finalize().into()
+}
+
+fn hash_pair(left: &[u8; 32], right: &[u8; 32]) -> [u8; 32] {
+    let mut ctx = Sha256::new();
+    ctx.update(&[2_u8]);
+    ctx.update(left);
+    ctx.update(right);
+    ctx.finalize().into()
+}
+
+// computes the tree-hash of a CLVM structure in serialized form
+pub fn tree_hash_from_stream(f: &mut Cursor<&[u8]>) -> io::Result<[u8; 32]> {
+    let mut values: Vec<[u8; 32]> = Vec::new();
+    let mut ops = vec![ParseOp::SExp];
+
+    let mut b = [0; 1];
+    loop {
+        let op = ops.pop();
+        if op.is_none() {
+            break;
+        }
+        match op.unwrap() {
+            ParseOp::SExp => {
+                f.read_exact(&mut b)?;
+                if b[0] == CONS_BOX_MARKER {
+                    ops.push(ParseOp::Cons);
+                    ops.push(ParseOp::SExp);
+                    ops.push(ParseOp::SExp);
+                } else if b[0] == 0x80 {
+                    values.push(hash_atom(&[]));
+                } else if b[0] <= MAX_SINGLE_BYTE {
+                    values.push(hash_atom(&b));
+                } else {
+                    let blob_size = decode_size(f, b[0])?;
+                    let blob = &f.get_ref()[f.position() as usize..];
+                    if (blob.len() as u64) < blob_size {
+                        return Err(bad_encoding());
+                    }
+                    f.set_position(f.position() + blob_size);
```

Comment:
You could use `.seek` here, possibly slightly more clear. But maybe not. Probably no speed difference.